db/index_changes.go

```diff
@@ -403,17 +422,105 @@ func (db *Database) initializeChannelFeeds(channelsSince channels.TimedSet, opti
+//
+// From Clock:
+// If vb < vb-B, seq = MaxUint
+// If vb = vb-B, seq = seq-b
+// if vb > vb-B, seq = SinceClock seq
+// To Clock:
+// If vb < vb-G, seq = MaxInt
+// If vb = vb-G, seq = seq-G
+// if vb > vb-G, seq = SinceClock
+
+func calculateBackfillRange(backfillPosition base.VbSequence, triggerPosition base.VbSequence, sinceClock base.SequenceClock) (fromClock, toClock base.SequenceClock) {
+
+	fromClock = base.NewSequenceClockImpl()
+	toClock = base.NewSequenceClockImpl()
+	MAX_SEQUENCE := uint64(math.MaxUint64)
+
+	for vbInt, _ := range sinceClock.Value() {
+		vbNo := uint16(vbInt)
+		var fromSeq, toSeq uint64
+
+		// Calculate from sequence for vbucket
+		if vbNo < backfillPosition.Vb {
+			fromSeq = MAX_SEQUENCE
+		} else if vbNo > backfillPosition.Vb {
+			fromSeq = sinceClock.GetSequence(vbNo)
+		} else { // vbNo == backfillPosition.Vb
+			fromSeq = backfillPosition.Seq
+		}
+		fromClock.SetSequence(vbNo, fromSeq)
+
+		// Calculate to sequence for vbucket
+		if vbNo < triggerPosition.Vb {
+			toSeq = MAX_SEQUENCE
+		} else if vbNo > triggerPosition.Vb {
+			toSeq = sinceClock.GetSequence(vbNo)
+		} else { // vbNo==triggerPosition.Vb
+			toSeq = triggerPosition.Seq
+		}
+		toClock.SetSequence(vbNo, toSeq)
+	}
+
+	return fromClock, toClock
+}
+
 // Creates a Go-channel of all the changes made on a channel.
 // Does NOT handle the Wait option. Does NOT check authorization.
 func (db *Database) vectorChangesFeed(channel string, options ChangesOptions) (<-chan *ChangeEntry, error) {
 	dbExpvars.Add("channelChangesFeeds", 1)
-	log, err := db.changeCache.GetChanges(channel, options)
-	base.LogTo("Changes+", "[changesFeed] Found %d changes for channel %s", len(log), channel)
-	if err != nil {
-		return nil, err
+	changeIndex, ok := db.changeCache.(*kvChangeIndex)
+	if !ok {
+		return nil, errors.New("Called vectorChangesFeed with non-index cache.")
 	}
 
-	if len(log) == 0 {
+	// If we're in backfill for this channel, we make one reader call for the backfill and one for non-backfill.  Without
+	// two requests, it's not possible to use limit to prevent a full index scan of some vbuckets while still preserving ordering.
+	// The second call should only be made if the first request doesn't return limit changes.
+	var backfillLog, log []*LogEntry
+	var err error
+
+	if options.Since.TriggeredByClock != nil {
+		// Changes feed is in backfill for this channel.
+		// Backfill position: (vb,seq) position in the backfill. e.g. [0,0] if we're just starting the backfill, [vb,seq] if we're midway through.
+		// Trigger position: (vb,seq) of the document that triggered this backfill (e.g. access granting doc or user doc)
+		backfillPosition := base.VbSequence{options.Since.vbNo, options.Since.Seq}
+		triggerPosition := base.VbSequence{options.Since.TriggeredByVbNo, options.Since.TriggeredBy}
+		backfillFrom, backfillTo := calculateBackfillRange(backfillPosition, triggerPosition, options.Since.Clock)
```

Comment:
This code might be easier to read as:

```
// Backfill position: (vb,seq) position in the backfill. e.g. [0,0] if we're just starting the backfill, [vb,seq] if we're midway through.
backfillPosition := base.VbSequence{options.Since.vbNo, options.Since.Seq}

// Trigger position: (vb,seq) of the document that triggered this backfill (e.g. access granting doc or user doc)
triggerPosition := base.VbSequence{options.Since.TriggeredByVbNo, options.Since.TriggeredBy}

```